zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -303,19 +277,126 @@ private boolean checkFourLetterWord(final Channel channel,
         }
     }
 
-    public void receiveMessage(ChannelBuffer message) {
+    /**
+     * Process incoming message. This should only be called from the event
+     * loop thread.
+     * @param buf the message bytes to process.
+     */
+    void processMessage(ByteBuf buf) {
+        assert channel.eventLoop().inEventLoop();
```

Comment:
nit: in our internal fork, we changed these asserts to if-throws (I forgot to upstream this change, will do it on master branch shortly). I don't think it matters that much since they've never fired. Anyway, what do you think about replacing this assert with:

```
        if (!channel.eventLoop().inEventLoop()) {
            throw new IllegalStateException(
                    "processMessage() called from non-EventLoop thread");
        }
```

Reply:
I suggest to stay in-sync with master now until that patch is not accepted. After that we can backport and release it easily.